lib/std/smartptrs.nim

```diff
@@ -0,0 +1,196 @@
+#    distribution, for details about the copyright.
+
+## C++11 like Smart pointers. It always uses the shared allocator.
+
+runnableExamples:
+  import std/isolation
+
+  block:
+    var a1: UniquePtr[float]
+    var a2 = newUniquePtr(isolate(0))
+
+    assert $a1 == "UniquePtr[float](nil)"
+    assert a1.isNil == true
+    assert $a2 == "UniquePtr[int](0)"
+    assert a2.isNil == false
+    assert a2[] == 0
+
+    # UniquePtr can't be copied but can be moved
+    let a3 = move a2 # a2 will be destroyed
+
+    assert $a2 == "UniquePtr[int](nil)"
+    assert a2.isNil == true
+
+    assert $a3 == "UniquePtr[int](0)"
+    assert a3.isNil == false
+    assert a3[] == 0
+
+  block:
+    var a1: SharedPtr[float]
+    let a2 = newSharedPtr(isolate(0))
+    let a3 = a2
+
+    assert $a1 == "SharedPtr[float](nil)"
+    assert a1.isNil == true
+    assert $a2 == "SharedPtr[int](0)"
+    assert a2.isNil == false
+    assert a2[] == 0
+    assert $a3 == "SharedPtr[int](0)"
+    assert a3.isNil == false
+    assert a3[] == 0
+
+  block:
+    var a1: ConstPtr[float]
+    let a2 = newConstPtr(isolate(0))
+    let a3 = a2
+
+    assert $a1 == "ConstPtr[float](nil)"
+    assert a1.isNil == true
+    assert $a2 == "ConstPtr[int](0)"
+    assert a2.isNil == false
+    assert a2[] == 0
+    assert $a3 == "ConstPtr[int](0)"
+    assert a3.isNil == false
+    assert a3[] == 0
+
+
+import std/isolation
+
+
+type
+  UniquePtr*[T] = object
+    ## Non copyable pointer to object T, exclusive ownership of the object.
+    val: ptr T
+
+proc `=destroy`*[T](p: var UniquePtr[T]) =
+  if p.val != nil:
+    `=destroy`(p.val[])
+    when compileOption("threads"):
+      deallocShared(p.val)
+    else:
+      dealloc(p.val)
+
+proc `=`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.error.}
+  ## The copy operation is disallowed for `UniquePtr`, `UniquePtr`
+  ## can only be moved.
+
+proc newUniquePtr*[T](val: sink Isolated[T]): UniquePtr[T] {.nodestroy.} =
+  ## Returns a unique pointer which has exclusive ownership of the object.
+  when compileOption("threads"):
+    result.val = cast[ptr T](allocShared(sizeof(T)))
+  else:
+    result.val = cast[ptr T](alloc(sizeof(T)))
+  # thanks to '.nodestroy' we don't have to use allocShared0 here.
+  # This is compiled into a copyMem operation, no need for a sink
+  # here either.
+  result.val[] = val.extract
+  # no destructor call for 'val: sink T' here either.
+
+converter convertUniquePtrToObj*[T](p: UniquePtr[T]): var T {.inline.} =
+  when compileOption("boundChecks"):
+    assert(p.val != nil, "deferencing nil unique pointer")
+  p.val[]
+
+proc isNil*[T](p: UniquePtr[T]): bool {.inline.} =
+  p.val == nil
+
+proc `[]`*[T](p: UniquePtr[T]): var T {.inline.} =
+  when compileOption("boundChecks"):
+    assert(p.val != nil, "deferencing nil unique pointer")
+  p.val[]
+
+proc `$`*[T](p: UniquePtr[T]): string {.inline.} =
+  if p.val == nil: "UniquePtr[" & $T & "](nil)"
+  else: "UniquePtr[" & $T & "](" & $p.val[] & ")"
```

Comment:
Too much ` & & & & & & ` here, maybe use `result.add`, and feels too big for `{.inline.}`.

Reply:
https://github.com/nim-lang/Nim/pull/17333/files#r597725651